File: graphs/valiparentheses.py
'''
LeetCode - problem 2267
'''

    def hasValidPath(self, grid: List[List[str]]) -> bool:  
        m = len(grid)
        n = len(grid[0])
        @lru_cache(maxsize=None)
        def hasValidPathInner(x, y, cnt):
            if x == m or y == n or cnt < 0:
                return False
            
            cnt += 1 if grid[x][y] == '(' else -1
            
            if x == m - 1 and y == n - 1 and not cnt:
                return True
            
            return hasValidPathInner(x + 1, y, cnt) or hasValidPathInner(x, y + 1, cnt)

        return hasValidPathInner(0, 0, 0)

File: graphs/valiparentheses.cpp
//LeetCode - problem 2267

int visited[100][100][101] = {};

bool hasValidPath(vector<vector<char>>& grid, int i = 0, int j = 0, int bal = 0) {
    int m = grid.size(), n = grid[0].size();
    bal += grid[i][j] == '(' ? 1 : -1;
    if (bal < 0 || bal > (m + n) / 2 || visited[i][j][bal])
        return false;
    visited[i][j][bal] = true;
    if (i == m - 1 && j == n - 1 && bal == 0)
        return true;
    if (i < m - 1 && hasValidPath(grid, i + 1, j, bal))
        return true;
    if (j < n - 1 && hasValidPath(grid, i, j + 1, bal))
        return true;
    return false;
}
